pkg/spartan/plugin.go

```diff
@@ -0,0 +1,170 @@
+	"github.com/containernetworking/cni/pkg/ns"
+	"github.com/containernetworking/cni/pkg/skel"
+	"github.com/containernetworking/cni/pkg/types/current"
+)
+
+func setupContainerVeth(netns, ifName string, mtu int, pr current.Result, spartanIPs []net.IPNet) (string, error) {
+	// The IPAM result will be something like IP=192.168.3.5/24,
+	// GW=192.168.3.1. What we want is really a point-to-point link but
+	// veth does not support IFF_POINTOPONT. So we set the veth
+	// interface to 192.168.3.5/32. Since the device netmask is set to
+	// /32, this would not add any routes to the main routing table.
+	// Therefore, in order to reach the spartan interfaces, we will have
+	// to explicitly set routes to the spartan interface through this
+	// device.
+
+	var hostVethName string
+
+	err := ns.WithNetNSPath(netns, func(hostNS ns.NetNS) error {
+		hostVeth, _, err := ip.SetupVeth(ifName, mtu, hostNS)
+		if err != nil {
+			return err
+		}
+
+		containerVeth, err := netlink.LinkByName(ifName)
+		if err != nil {
+			return fmt.Errorf("failed to lookup container VETH %q: %s", ifName, err)
+		}
+
+		// Configure the container veth with IP address returned by the
+		// IPAM, but set the netmask to a /32.
+		if err := netlink.LinkSetUp(containerVeth); err != nil {
+			return fmt.Errorf("failed to set %q UP: %s", ifName, err)
+		}
+
+		// Set the netmask to a /32.
+		pr.IPs[0].Address.Mask = net.IPv4Mask(0xff, 0xff, 0xff, 0xff)
+
+		addr := &netlink.Addr{IPNet: &pr.IPs[0].Address, Label: ""}
+		if err = netlink.AddrAdd(containerVeth, addr); err != nil {
+			return fmt.Errorf("failed to add IP address to %q: %s", ifName, err)
+		}
+
+		// Add routes to the spartan interfaces through this interface.
+		for _, spartanIP := range spartanIPs {
+			spartanRoute := netlink.Route{
+				LinkIndex: containerVeth.Attrs().Index,
+				Dst:       &spartanIP,
+				Scope:     netlink.SCOPE_LINK,
+				Src:       pr.IPs[0].Address.IP,
+			}
+
+			if err = netlink.RouteAdd(&spartanRoute); err != nil {
+				return fmt.Errorf("failed to add spartan route %s: %s", spartanRoute, err)
+			}
+		}
+
+		hostVethName = hostVeth.Name
+
+		return nil
+	})
+
+	return hostVethName, err
+}
+
+func CniAdd(args *skel.CmdArgs) error {
+	// Delegate plugin seems to be successful, install the spartan
+	// network.
+	spartanNetConf, err := json.Marshal(Config)
+	if err != nil {
+		return fmt.Errorf("failed to marshall the `spartan-network` IPAM configuration: %s", err)
+	}
+
+	// Run the IPAM plugin for the spartan network.
+	ipamResult, err := ipam.ExecAdd(Config.IPAM.Type, spartanNetConf)
+	if err != nil {
+		return err
+	}
+
+	result, err := current.NewResultFromResult(ipamResult)
+	if err != nil {
+		return err
+	}
+
+	if result.IPs == nil {
+		return errors.New("IPAM plugin returned missing IPv4 config")
+	}
+
+	// Make sure we got only one IP and that it is IPv4
+	switch {
+	case len(result.IPs) > 1:
+		return errors.New("Expecting a single IPv4 address from IPAM")
```

Comment:
isn't there a current discussion happening elsewhere for ipv6 support? if so, should we take inv6 into account here?